File: app/src/androidTest/java/uk/co/victoriajanedavis/chatapp/test_common/ModelGenerationUtil.java
package uk.co.victoriajanedavis.chatapp.test_common;

import java.util.ArrayList;
import java.util.List;
import java.util.UUID;

import io.reactivex.annotations.Nullable;
import uk.co.victoriajanedavis.chatapp.data.model.db.ChatDbModel;
import uk.co.victoriajanedavis.chatapp.data.model.db.FriendshipDbModel;
import uk.co.victoriajanedavis.chatapp.data.model.network.*;

public class ModelGenerationUtil {

    public static List<ChatMembershipNwModel> createChatMembershipNwList(int number) {
        List<ChatMembershipNwModel> chatNwModels = new ArrayList<>(number);
        for (int i=0; i<number; i++) {
            chatNwModels.add(new ChatMembershipNwModel(createChatNwModel(), createUserNwModel(null)));
        }

        return chatNwModels;
    }

    public static ChatNwModel createChatNwModel() {
        return new ChatNwModel(UUID.randomUUID());
    }

    public static UserNwModel createUserNwModel(@Nullable String username) {
        UserNwModel model = new UserNwModel(UUID.randomUUID());
        model.setUsername(username);
        return model;
    }

    public static List<UserNwModel> createUserNwList(int number) {
        List<UserNwModel> userNwModels = new ArrayList<>(number);
        for (int i=0; i<number; i++) {
            userNwModels.add(createUserNwModel(null));
        }

        return userNwModels;
    }

    public static List<ChatDbModel> createChatMembershipDbModelList(int number) {
        List<ChatDbModel> chatDbModelList = new ArrayList<>(number);
        for (int i=0; i<number; i++) {
            chatDbModelList.add(createChatMembershipDbModelWithFriendshipDbModel());
        }

        return chatDbModelList;
    }

    public static ChatDbModel createChatMembershipDbModelWithFriendshipDbModel() {
        ChatDbModel dbModel = createChatMembershipDbModel();

        dbModel.setFriendship(createFriendshipDbModel(dbModel));
        return dbModel;
    }

    public static ChatDbModel createChatMembershipDbModel() {
        ChatDbModel dbModel = new ChatDbModel();
        dbModel.setUuid(UUID.randomUUID());

        return dbModel;
    }

    public static FriendshipDbModel createFriendshipDbModel(ChatDbModel chatDbModel) {
        FriendshipDbModel friendshipDbModel = new FriendshipDbModel();
        friendshipDbModel.setUuid(UUID.randomUUID());

        friendshipDbModel.setChatUuid(chatDbModel.getUuid());
        return friendshipDbModel;
    }

    public static List<MessageNwModel> createMessageNwList(UUID chatUuid, int number) {
        List<MessageNwModel> messageNwModels = new ArrayList<>(number);
        for (int i=0; i<number; i++) {
            messageNwModels.add(createMessageNwModel(chatUuid));
        }

        return messageNwModels;
    }

    public static MessageNwModel createMessageNwModel(@Nullable UUID chatUuid) {
        MessageNwModel model = new MessageNwModel(UUID.randomUUID());
        model.setChatUuid(chatUuid);
        //model.setUser(createUserNwModel(null));
        return model;
    }

    public static TokenNwModel createTokenNwModel() {
        TokenNwModel model = new TokenNwModel();

        model.setToken(UUID.randomUUID().toString());
        model.setUser(createUserNwModel(null));
        return model;
    }
}

File: app/src/test/java/uk/co/victoriajanedavis/chatapp/common/ModelGenerationUtil.kt
package uk.co.victoriajanedavis.chatapp.common

import io.reactivex.annotations.Nullable
import uk.co.victoriajanedavis.chatapp.data.common.TimestampProvider
import uk.co.victoriajanedavis.chatapp.data.model.db.ChatDbModel
import uk.co.victoriajanedavis.chatapp.data.model.db.FriendshipDbModel
import uk.co.victoriajanedavis.chatapp.data.model.db.MessageDbModel
import uk.co.victoriajanedavis.chatapp.data.model.network.*
import java.util.*

object ModelGenerationUtil {

    /*
    fun createChatMembershipNwList(number: Int): List<ChatMembershipNwModel> {
        val chatNwModels = ArrayList<ChatMembershipNwModel>(number)
        for (i in 0 until number) {
            chatNwModels.add(ChatMembershipNwModel(createChatNwModel(), createUserNwModel(null)))
        }

        return chatNwModels
    }


    fun createChatNwModel(): ChatNwModel {
        return ChatNwModel(UUID.randomUUID())
    }

    fun createUserNwModel(@Nullable username: String?): UserNwModel {
        val model = UserNwModel(UUID.randomUUID())
        model.setUsername(username)
        return model
    }

    fun createUserNwList(number: Int): List<UserNwModel> {
        val userNwModels = ArrayList<UserNwModel>(number)
        for (i in 0 until number) {
            userNwModels.add(createUserNwModel(null))
        }

        return userNwModels
    }

    fun createChatMembershipDbModelList(number: Int): List<ChatDbModel> {
        val chatDbModelList = ArrayList<ChatDbModel>(number)
        for (i in 0 until number) {
            chatDbModelList.add(createChatMembershipDbModelWithFriendshipDbModel())
        }

        return chatDbModelList
    }

    fun createChatMembershipDbModelWithFriendshipDbModel(): ChatDbModel {
        val dbModel = createChatMembershipDbModel()

        dbModel.setFriendship(createFriendshipDbModel(dbModel))
        return dbModel
    }

    */

    @JvmStatic
    fun createMessageDbModel(
        chatUuid: UUID = UUID.randomUUID(),
        text: String = "",
        created: Date = Date(TimestampProvider.currentTimeMillis())
    ): MessageDbModel {
        return MessageDbModel(
            uuid = UUID.randomUUID(),
            text = text,
            created = created,
            chatUuid = chatUuid,
            userUuid = UUID.randomUUID(),
            userUsername = "Username",
            isFromCurrentUser = true
        )
    }

    /*
    fun createFriendshipDbModel(chatDbModel: ChatDbModel): FriendshipDbModel {
        val friendshipDbModel = FriendshipDbModel()
        friendshipDbModel.uuid = UUID.randomUUID()

        friendshipDbModel.chatUuid = chatDbModel.getUuid()
        return friendshipDbModel
    }
    */

    @JvmStatic
    fun createMessageNwList(chatUuid: UUID, number: Int): List<MessageNwModel> {
        return MutableList(number) { createMessageNwModel(chatUuid = chatUuid) }
    }

    @JvmStatic
    fun createMessageNwModel(
        chatUuid: UUID = UUID.randomUUID(),
        created: Date = Date(TimestampProvider.currentTimeMillis())
    ): MessageNwModel {
        return MessageNwModel(
            uuid = UUID.randomUUID(),
            text = "",
            created = created,
            chatUuid = chatUuid,
            userUuid = UUID.randomUUID(),
            userUsername = "Username",
            isFromCurrentUser = true
        )
    }

    /*
    fun createTokenNwModel(): TokenNwModel {
        val model = TokenNwModel()

        model.setToken(UUID.randomUUID().toString())
        model.setUser(createUserNwModel(null))
        return model
    }
    */
}
